### PR TITLE
resolve #18

### PR DESCRIPTION
**Pull request by Guseyn**

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#21](https://github.com/artipie/pypi-adapter/issues/21) is already in scope

**Comment by 0crat** [comment]
Job `gh:artipie/pypi-adapter#21` is not assigned, can't get performer

**Comment by g4s8** [comment]
@0crat assign @amihaiemil

**Comment by 0crat** [comment]
Pull request [#21](https://github.com/artipie/pypi-adapter/issues/21) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

**Comment by 0crat** [comment]
> @0crat assign @amihaiemil  ([here](https://github.com/artipie/pypi-adapter/issues/21#issuecomment-584527949))

@g4s8 This pull request [#21](https://github.com/artipie/pypi-adapter/issues/21) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/7df32fe1-3953-4be5-9f8d-353f2e8867ce); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by amihaiemil** [comment]
@Guseyn I'm looking over this now. In the meantime, add a Description and link the proper Issue that this PR is solving. There is no Issue 18, you probably got it wrong.

File: src/main/java/com/artipie/pypi/Artifact.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+public interface Artifact {
+    /**
+     * Saves itself into abstract storage.
+     *
+     * @param storage Storage which is for saving the artifact.
+     * @return Future of save operation.
+     */
+    CompletableFuture<Void> save(Storage storage);
+}
```
Comment (amihaiemil) [inline src/main/java/com/artipie/pypi/Artifact.java]: @Guseyn Didn't you already introduce this file in a previous PR? 

Also, I remember I suggested that the ``save`` method should return the ``Key`` to which this Artifact has been mapped in the Storage.

File: src/main/java/com/artipie/pypi/ArtifactMeta.java [hunk 2]
```diff
@@ -49,27 +44,24 @@
...
+    public ArtifactMeta(final String name, final String pattern) {
         this.name = name;
-        this.version = version;
         this.pattern = pattern;
     }
 
     @Override
     public String html() {
-        return String.format(this.pattern, this.name, this.version, this.name, this.version);
+        return String.format(this.pattern, this.name, this.name);
```
Comment (amihaiemil) [inline src/main/java/com/artipie/pypi/ArtifactMeta.java]: @Guseyn you're adding the name twice to the format? Why?

File: src/main/java/com/artipie/pypi/ArtifactMeta.java [hunk 3]
```diff
@@ -49,27 +44,24 @@
      * Ctor.
      *
      * @param name Name of artifact.
-     * @param version Version of artifact.
      */
-    public ArtifactMeta(final String name, final String version) {
```
Comment (amihaiemil) [inline src/main/java/com/artipie/pypi/ArtifactMeta.java]: @Guseyn Why remove the version?

File: src/main/java/com/artipie/pypi/BinaryArtifact.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    /**
+     * Ctor.
+     *
+     * @param name Name of artifact.
+     * @param content Content of artifact.
+     */
+    public BinaryArtifact(final String name, final String content) {
+        this.name = name;
+        this.content = new ByteFlow(content);
+    }
```
Comment (amihaiemil) [inline src/main/java/com/artipie/pypi/BinaryArtifact.java]: @Guseyn Same discussion as in the previous PR. This constructor should accept ``Flow<ByteBuffer>``, not ``String`` content. If the ``content`` is a ``String``, then this class becomes ``StringBinaryArtifact``. 

Instead, you should have a central ``Flow<ByteBuffer>`` constructor, and some other, cascading constructors. See:

```java
    ...
    /**
     * Cascade constructor so we can  build this BinaryArtifact from a String.
     *
     * @param name Name of artifact.
     * @param content Content of artifact.
     */
    public BinaryArtifact(final String name, final String content) {
        this(name, new ByteFlow(content));
    }

   /**
     * Initializing constructor.
     *
     * @param name Name of artifact.
     * @param content Content of artifact.
     */
    public BinaryArtifact(final String name, final Flow<ByteBuffer> content) {
        this.name = name;
        this.content = content;
    }
   ...
```

File: src/main/java/com/artipie/pypi/ByteFlow.java [hunk 5]
```diff
@@ -34,34 +35,39 @@
  *
  * @since 0.1
  */
-public final class ByteFlow implements Flow<Byte> {
+public final class ByteFlow implements Flow<ByteBuffer> {
```
Comment (amihaiemil) [inline src/main/java/com/artipie/pypi/ByteFlow.java]: @Guseyn Why did you change from Byte to ByteBuffer?

File: src/test/java/com/artipie/pypi/ByteFlowTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ * ByteFlowTest.
+ *
+ * @since 0.1
+ */
+public class ByteFlowTest {
+    /**
+     * To string test.
+     */
+    @Test
+    public void string() {
```
Comment (amihaiemil) [inline src/test/java/com/artipie/pypi/ByteFlowTest.java]: @Guseyn Specify a better name. For instance ``toStringWorks()``...

File: src/main/java/com/artipie/pypi/BinaryArtifact.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    public CompletableFuture<Void> save(final Storage storage) {
+        final String pkgname = this.name.split("-\\d.\\d.\\d.tar.gz")[0];
+        return storage.save(new Key.From(pkgname, this.name), this.content.value());
+    }
+}
```
Comment (amihaiemil) [inline src/main/java/com/artipie/pypi/BinaryArtifact.java]: @Guseyn Some tests for this class are needed, I think :)

File: src/main/java/com/artipie/pypi/Artifact.java (hunk 1 above)
Comment (Guseyn) [reply src/main/java/com/artipie/pypi/Artifact.java]: @amihaiemil again, I use `asto` from artipie `project`, so we have to return CompletableFuture from save operation and we have to pass `key` param, we are not supposed to return it? Why do you think that `Key` is something that we need to return? It works like `key-value` storage, like method `put` in `Map` interface

File: src/main/java/com/artipie/pypi/ArtifactMeta.java (hunk 2 above)
Comment (Guseyn) [reply src/main/java/com/artipie/pypi/ArtifactMeta.java]: @amihaiemil because `pattern` contains name twice. Read the code before asking such questions.

File: src/main/java/com/artipie/pypi/ArtifactMeta.java (hunk 3 above)
Comment (Guseyn) [reply src/main/java/com/artipie/pypi/ArtifactMeta.java]: @amihaiemil Because version comes in the `name`, and it's a part of the name, so we don't need it separately.

File: src/main/java/com/artipie/pypi/ArtifactMeta.java (hunk 2 above)
Comment (amihaiemil) [reply src/main/java/com/artipie/pypi/ArtifactMeta.java]: @Guseyn But maybe it's enough to say ``String.format(pattern, name)`` and the name will go in all ``%s`` from the pattern? I'm not sure though.

File: src/main/java/com/artipie/pypi/Artifact.java (hunk 1 above)
Comment (amihaiemil) [reply src/main/java/com/artipie/pypi/Artifact.java]: @Guseyn Depends. 

If you say ``map.put(key, value)`` then yes, you are right. But if I say ``map.pus(value)``, it means that the map will generate the key for me. And once that key is generated, it should be returned to me, right? Otherwise how will I be able to fetch the value from the map later?

And in this case, you have ``map.save(value)``. 

Makes sense?

File: src/main/java/com/artipie/pypi/BinaryArtifact.java (hunk 4 above)
Comment (Guseyn) [reply src/main/java/com/artipie/pypi/BinaryArtifact.java]: @amihaiemil BinaryArtifact needs only constructor with string otherwise it would be over engineering. If I would need constructor with bytes, I'll add it. And I don't add more unnecessary classes not only because it will increase the size of changes in pull request, but also because it's  unnecessary.

File: src/main/java/com/artipie/pypi/ByteFlow.java (hunk 5 above)
Comment (Guseyn) [reply src/main/java/com/artipie/pypi/ByteFlow.java]: @amihaiemil as I told you in the prev. pr `asto` has changed its behaviour, not it's more efficient because we use **ByteBuffer**, it allows us read something by chucks of bytes, not just by each byte.

File: src/main/java/com/artipie/pypi/BinaryArtifact.java (hunk 4 above)
Comment (amihaiemil) [reply src/main/java/com/artipie/pypi/BinaryArtifact.java]: @Guseyn You don't understand. It's not about necessary/unnecessary, it's about correct/incorrect.

At the moment your class is incorrect, there is a conflict between its name and its constructor. The name suggests that it's a general-purpose BinaryArtifact, while the constructor only lets you build it from String, which means the name should actually be StringBinaryAratifact!

**Comment by amihaiemil** [comment]
@Guseyn Please see my replies :)

File: src/test/java/com/artipie/pypi/ByteFlowTest.java (hunk 6 above)
Comment (Guseyn) [reply src/test/java/com/artipie/pypi/ByteFlowTest.java]: @amihaiemil ok, I'll change it.

File: src/main/java/com/artipie/pypi/BinaryArtifact.java (hunk 7 above)
Comment (Guseyn) [reply src/main/java/com/artipie/pypi/BinaryArtifact.java]: @amihaiemil I am going add integration tests in another pr, as it would take some time and amount of change line, otherwise this pr would be much bigger.

**Comment by Guseyn** [comment]
@amihaiemil please see the updates and replies above

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by amihaiemil** [comment]
@Guseyn Thanks, I read your replies and I ansewered to 3 of them, please check.

Also, why don't you want to specify a Description and an Issue to this PR?! 
If I were the Architect or Project Owner I would close this PR instantly

"resolve #18" -- there is no ticket 18 and what do you mean by "resolve" anyway.
* not Description * -- what is this PR, why does it have no description? 

Please be more disciplined, even if you are the only DEV on this project right now. We need traceablity of changes. There cannot be any traceability if you don't describe and sync PRs with Issues!

File: src/main/java/com/artipie/pypi/BinaryArtifact.java (hunk 4 above)
Comment (Guseyn) [reply src/main/java/com/artipie/pypi/BinaryArtifact.java]: @amihaiemil  No, it's correct as it accepts content as a string but it contains it as a `ByteFlow`.
It can potentially have constructor that accepts other types of input, but at the end it'll be converted in the `ByteFlow`. Param `content` is `ByteFlow` and it does not matter from what kind of input param it is constructed.
```
 this.content = new ByteFlow(content);
```
It's `ByteFlow` anyway

Comment (amihaiemil) [reply src/main/java/com/artipie/pypi/BinaryArtifact.java]: @Guseyn And what I pull this project as a dependency to my project and I want to use this class? It's compiled code, I cannot edit it to add constructors. I am told that it's a BinaryArtifact but I can only build it from String. 

First of all, personally, I would expect to build this object from a ``Path`` (path to the artifact on disk). 

Anyway. Let's leave it as you think....

**Comment by Guseyn** [comment]
@amihaiemil "Also, why don't you want to specify a Description and an Issue to this PR?!" => because I can write "resolve #18", which means that you can click on #18 and see what kind of things I am trying to resolve by this pr. The key word `resolve` in the pr automatically closes issue with number `18`.

File: src/main/java/com/artipie/pypi/BinaryArtifact.java (hunk 4 above)
Comment (Guseyn) [reply src/main/java/com/artipie/pypi/BinaryArtifact.java]: We work with abstract storage `asto`, so we don't have such thing as `Path`, because we have `Key` which is constructed by name or several names.

File: src/main/java/com/artipie/pypi/ArtifactMeta.java (hunk 2 above)
Comment (Guseyn) [reply src/main/java/com/artipie/pypi/ArtifactMeta.java]: @amihaiemil No it doesn't work like that.

**Comment by Guseyn** [comment]
@amihaiemil see the replies above.

File: src/main/java/com/artipie/pypi/Artifact.java (hunk 1 above)
Comment (Guseyn) [reply src/main/java/com/artipie/pypi/Artifact.java]: @amihaiemil As I told you, I just use `asto` and I have to follow its behaviour. We don't have to generate `Key`, we must control it.

**Comment by amihaiemil** [comment]
@Guseyn 

> because I can write "resolve #18", which means that you can click on #18 

You cannot click on the number if it's just in the PR Title! And in the case of this PR the Issue number is not correct! #18 is not an Issue, it is a previous Pull Request. 

Furthermore, it helps to Describe the PR and clearly state what the changes are. Someone might see this PR and might not have the time to dig through the files.

I can't believe it's so hard to understand.

**Comment by 0crat** [comment]
Job was finished in 4 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Order was finished: +20 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)
